conda/core/index.py

```diff
@@ -51,7 +58,566 @@ def check_allowlist(channel_urls: list[str]) -> None:
+                pass
+        if self.track_features and key.name.endswith("@"):
+            for track_feature in self.track_features:
+                if track_feature == key.name[:-1]:
+                    return PackageRecord.feature(track_feature)
+        prec = self._retrieve_from_channels(key)
+        prec = self._update_from_prefix(key, prec)
+        if self.use_cache:
+            prec = self._update_from_cache(key, prec)
+        if prec is None:
+            raise KeyError((key,))
+        return prec
+
+    def __contains__(self, key: PackageRecord) -> bool:
+        try:
+            _ = self[key]
+            return True
+        except (PackagesNotFoundError, KeyError):
+            return False
+
+    def __copy__(self) -> Self:
+        inst = self.__class__.__new__(self.__class__)
+        inst.__dict__.update(self.__dict__)
+        if "_data" in self.__dict__:
+            inst.__dict__["_data"] = self.__dict__["_data"].copy()
+        return inst
+
+
+class ReducedIndex(Index):
+    """Index that contains a subset of available packages.
+
+    Like :class:`Index`, this makes information about packages from the same four
+    sources available. However, the contents of the reduced index is limited to
+    a subset of packages relevant to a given specification.
+    This works by taking into account all packages that match the given specification
+    together with their dependencies and their dependencies dependencies, etc.
+
+    Note:
+        See :meth:`Index.get_reduced_index` for convenient construction.
+    """
+
+    def __init__(
+        self,
+        specs: Iterable[MatchSpec],
+        channels: tuple[str, ...] = (),
+        prepend: bool = True,
+        platform: str | None = None,
+        subdirs: tuple[str, ...] | None = None,
+        use_local: bool = False,
+        use_cache: bool | None = None,
+        prefix: str | None = None,
+        repodata_fn: str | None = context.repodata_fns[-1],
+        use_system: bool = False,
+    ) -> None:
+        """Initialize a new reduced index.
+
+        Args:
+          specs: the collection of specifications that span the subset of packages.
+          all other args: see :class:`Index`.
+        """
+        super().__init__(
+            channels,
+            prepend,
+            platform,
+            subdirs,
+            use_local,
+            use_cache,
+            prefix,
+            repodata_fn,
+            use_system,
+        )
+        self.specs = specs
+        self._derive_reduced_index()
+
+    def __repr__(self) -> str:
+        channels = ", ".join(self.channels.keys())
+        return f"<ReducedIndex(spec={self.specs}, channels=[{channels}])>"
+
+    def _derive_reduced_index(self) -> None:
+        records = IndexedSet()
+        collected_names = set()
+        collected_track_features = set()
+        pending_names = set()
+        pending_track_features = set()
+
+        def push_spec(spec: MatchSpec) -> None:
```

Comment:
`push_spec` is called several times below, often within a loop, thoughts on updating it to support `*specs`?

```python
def push_specs(*specs: MatchSpec | str) -> None:
    for spec in map(MatchSpec, specs):
        ...
```

that way instead of writing:

```python
push_spec(MatchSpec(record.name))
for _spec in combined_depends:
    push_spec(_spec)
if record.track_features:
    for ftr_name in record.track_features:
        push_spec(MatchSpec(track_features=ftr_name))
```

we can write:

```python
push_specs(record.name, *combined_deps, *record.track_features)
```

looks like `push_record` could benefit from similar changes

Reply:
Nice and done, thanks!